File: src/ios/BabyCryAnalyzer/Background/BackgroundTaskModule.m
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>
#import <AVFoundation/AVFoundation.h>
#import <BackgroundTasks/BackgroundTasks.h>

// Error domain for module-specific errors
static NSString *const ErrorDomain = @"com.babycryanalyzer.backgroundtask";
static NSString *const BGTaskIdentifier = @"com.babycryanalyzer.monitoring";

// Error codes
typedef NS_ENUM(NSInteger, BackgroundTaskError) {
    BackgroundTaskErrorInvalidState = 1001,
    BackgroundTaskErrorAudioSessionFailed = 1002,
    BackgroundTaskErrorSchedulingFailed = 1003,
    BackgroundTaskErrorInvalidConfiguration = 1004
};

@interface BackgroundTaskModule : RCTEventEmitter <RCTBridgeModule>

@property (nonatomic, strong) AVAudioSession *audioSession;
@property (nonatomic, strong) BGTaskScheduler *taskScheduler;
@property (nonatomic, assign) BOOL isMonitoring;
@property (nonatomic, strong) NSDictionary *currentConfig;

@end

@implementation BackgroundTaskModule

RCT_EXPORT_MODULE()

- (instancetype)init {
    if (self = [super init]) {
        _audioSession = [AVAudioSession sharedInstance];
        _taskScheduler = [BGTaskScheduler sharedScheduler];
        _isMonitoring = NO;
        _currentConfig = @{
            @"monitoringInterval": @900, // 15 minutes default
            @"sensitivity": @"medium",
            @"noiseThreshold": @0.3
        };
        
        // Register for audio session interruption notifications
        [[NSNotificationCenter defaultCenter] addObserver:self
                                               selector:@selector(handleAudioSessionInterruption:)
                                                   name:AVAudioSessionInterruptionNotification
                                                 object:nil];
    }
    return self;
}

- (NSArray<NSString *> *)supportedEvents {
    return @[@"monitoringStateChanged", @"backgroundTaskScheduled", @"audioSessionError"];
}

- (BOOL)initAudioSession {
    NSError *error = nil;
    
    // Configure audio session for background recording
    if (![_audioSession setCategory:AVAudioSessionCategoryRecord
                            mode:AVAudioSessionModeMeasurement
                         options:AVAudioSessionCategoryOptionAllowBluetooth
                           error:&error]) {
        return NO;
    }
    
    // Set active
    if (![_audioSession setActive:YES error:&error]) {
        return NO;
    }
    
    return YES;
}

- (void)scheduleBackgroundTask:(NSDictionary *)config {
    // Register background task
    [[BGTaskScheduler sharedScheduler] registerForTaskWithIdentifier:BGTaskIdentifier
                                                        usingQueue:dispatch_get_main_queue()
                                                     launchHandler:^(__kindof BGTask * _Nonnull task) {
        [self handleBackgroundTask:task];
    }];
    
    BGAppRefreshTaskRequest *request = [[BGAppRefreshTaskRequest alloc] initWithIdentifier:BGTaskIdentifier];
    NSTimeInterval interval = [config[@"monitoringInterval"] doubleValue];
    request.earliestBeginDate = [NSDate dateWithTimeIntervalSinceNow:interval];
    
    NSError *error = nil;
    if (![[BGTaskScheduler sharedScheduler] submitTaskRequest:request error:&error]) {
        [self sendEventWithName:@"backgroundTaskScheduled" body:@{@"success": @NO, @"error": error.localizedDescription}];
        return;
    }
    
    [self sendEventWithName:@"backgroundTaskScheduled" body:@{@"success": @YES}];
}

- (void)handleBackgroundTask:(BGTask *)task {
    // Ensure we reschedule before processing
    [self scheduleBackgroundTask:self.currentConfig];
    
    // Process audio in background
    [self processAudioInBackground:^(BOOL success, NSError *error) {
        if (success) {
            [task setTaskCompletedWithSuccess:YES];
        } else {
            [task setTaskCompletedWithSuccess:NO];
        }
    }];
}

- (void)processAudioInBackground:(void (^)(BOOL success, NSError *error))completion {
    // Implementation of audio processing logic
    // This would integrate with the audio analysis service
    completion(YES, nil);
}

RCT_EXPORT_METHOD(startBackgroundMonitoring:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    if (self.isMonitoring) {
        reject(@"invalid_state", @"Background monitoring is already active", nil);
        return;
    }
    
    if (![self initAudioSession]) {
        NSError *error = [NSError errorWithDomain:ErrorDomain
                                           code:BackgroundTaskErrorAudioSessionFailed
                                       userInfo:@{NSLocalizedDescriptionKey: @"Failed to initialize audio session"}];
        reject(@"audio_session_failed", error.localizedDescription, error);
        return;
    }
    
    [self scheduleBackgroundTask:self.currentConfig];
    self.isMonitoring = YES;
    
    [self sendEventWithName:@"monitoringStateChanged" body:@{@"isMonitoring": @YES}];
    resolve(@{@"success": @YES});
}

RCT_EXPORT_METHOD(stopBackgroundMonitoring:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    if (!self.isMonitoring) {
        reject(@"invalid_state", @"Background monitoring is not active", nil);
        return;
    }
    
    [[BGTaskScheduler sharedScheduler] cancelAllTaskRequests];
    
    NSError *error = nil;
    if (![self.audioSession setActive:NO error:&error]) {
        reject(@"audio_session_failed", @"Failed to deactivate audio session", error);
        return;
    }
    
    self.isMonitoring = NO;
    [self sendEventWithName:@"monitoringStateChanged" body:@{@"isMonitoring": @NO}];
    resolve(@{@"success": @YES});
}

RCT_EXPORT_METHOD(getMonitoringStatus:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    resolve(@{
        @"isMonitoring": @(self.isMonitoring),
        @"currentConfig": self.currentConfig
    });
}

RCT_EXPORT_METHOD(updateConfiguration:(NSDictionary *)config
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    if (![self validateConfiguration:config]) {
        NSError *error = [NSError errorWithDomain:ErrorDomain
                                           code:BackgroundTaskErrorInvalidConfiguration
                                       userInfo:@{NSLocalizedDescriptionKey: @"Invalid configuration provided"}];
        reject(@"invalid_config", error.localizedDescription, error);
        return;
    }
    
    self.currentConfig = [self.currentConfig mutableCopy];
    [self.currentConfig addEntriesFromDictionary:config];
    
    if (self.isMonitoring) {
        [self scheduleBackgroundTask:self.currentConfig];
    }
    
    resolve(@{@"success": @YES});
}

- (BOOL)validateConfiguration:(NSDictionary *)config {
    // Validate configuration parameters
    if (config[@"monitoringInterval"]) {
        NSNumber *interval = config[@"monitoringInterval"];
        if ([interval doubleValue] < 900 || [interval doubleValue] > 3600) {
            return NO;
        }
    }
    
    if (config[@"sensitivity"]) {
        NSString *sensitivity = config[@"sensitivity"];
        if (![@[@"low", @"medium", @"high"] containsObject:sensitivity]) {
            return NO;
        }
    }
    
    if (config[@"noiseThreshold"]) {
        NSNumber *threshold = config[@"noiseThreshold"];
        if ([threshold doubleValue] < 0 || [threshold doubleValue] > 1) {
            return NO;
        }
    }
    
    return YES;
}

- (void)handleAudioSessionInterruption:(NSNotification *)notification {
    NSInteger type = [notification.userInfo[AVAudioSessionInterruptionTypeKey] integerValue];
    
    if (type == AVAudioSessionInterruptionTypeBegan) {
        [self sendEventWithName:@"audioSessionError" 
                          body:@{@"type": @"interruption", @"message": @"Audio session interrupted"}];
    } else if (type == AVAudioSessionInterruptionTypeEnded) {
        if (self.isMonitoring) {
            [self initAudioSession];
        }
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

File: src/ios/BabyCryAnalyzer/Background/BackgroundTaskModule.swift
// Foundation framework - iOS 14.0+
import Foundation
import React

// MARK: - Constants
private let kModuleName = "BackgroundTaskModule"
private let kSerialQueue = DispatchQueue(label: "com.babycryanalyzer.background")
private let kErrorDomain = "com.babycryanalyzer.background.error"

// MARK: - BackgroundTaskModule
@objc(BackgroundTaskModule)
public class BackgroundTaskModule: RCTEventEmitter {
    
    // MARK: - Properties
    private let monitoringService: BackgroundMonitoringService
    private let configuration: BackgroundConfiguration
    private let serialQueue: DispatchQueue
    private let configLock: NSLock
    
    // MARK: - Initialization
    override init() {
        self.configuration = BackgroundConfiguration()
        self.serialQueue = kSerialQueue
        self.configLock = NSLock()
        
        // Initialize monitoring service with default configuration
        let audioConfig = AudioConfiguration()
        let audioProcessor = try? AudioProcessor(config: audioConfig)
        self.monitoringService = BackgroundMonitoringService(config: configuration, processor: audioProcessor!)
        
        super.init()
        
        // Set up monitoring service delegate
        monitoringService.delegate = self
        
        // Register for battery notifications
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleBatteryLevelChange),
            name: UIDevice.batteryLevelDidChangeNotification,
            object: nil
        )
        UIDevice.current.isBatteryMonitoringEnabled = true
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    // MARK: - RCTEventEmitter Override
    override public static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override public func supportedEvents() -> [String] {
        return [
            "onMonitoringStateChange",
            "onCryDetected",
            "onBatteryOptimizationChange",
            "onError"
        ]
    }
    
    // MARK: - Public Methods
    @objc(startBackgroundMonitoring:withRejecter:)
    public func startBackgroundMonitoring(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        configLock.lock()
        defer { configLock.unlock() }
        
        guard configuration.isEnabled else {
            reject(
                "CONFIG_DISABLED",
                "Background monitoring is disabled in configuration",
                NSError(domain: kErrorDomain, code: -1)
            )
            return
        }
        
        serialQueue.async { [weak self] in
            guard let self = self else { return }
            
            switch self.monitoringService.startMonitoring() {
            case .success:
                self.sendEvent(withName: "onMonitoringStateChange", body: ["isMonitoring": true])
                resolve(["status": "started"])
                
            case .failure(let error):
                let nsError = NSError(domain: kErrorDomain, code: -2, userInfo: [
                    NSLocalizedDescriptionKey: error.localizedDescription
                ])
                reject("START_FAILED", error.localizedDescription, nsError)
            }
        }
    }
    
    @objc(stopBackgroundMonitoring:withRejecter:)
    public func stopBackgroundMonitoring(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        serialQueue.async { [weak self] in
            guard let self = self else { return }
            
            self.monitoringService.stopMonitoring()
            self.sendEvent(withName: "onMonitoringStateChange", body: ["isMonitoring": false])
            resolve(["status": "stopped"])
        }
    }
    
    @objc(getMonitoringStatus:withRejecter:)
    public func getMonitoringStatus(
        _ resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        serialQueue.async { [weak self] in
            guard let self = self else { return }
            
            let isMonitoring = self.monitoringService.isMonitoring
            let batteryImpact = self.monitoringService.getCurrentBatteryImpact()
            
            let status: [String: Any] = [
                "isMonitoring": isMonitoring,
                "batteryOptimizationEnabled": self.configuration.batteryOptimizationEnabled,
                "batteryImpact": batteryImpact,
                "monitoringInterval": self.configuration.monitoringInterval,
                "minimumBatteryLevel": self.configuration.minimumBatteryLevel
            ]
            
            resolve(status)
        }
    }
    
    @objc(updateConfiguration:withResolver:withRejecter:)
    public func updateConfiguration(
        _ config: NSDictionary,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        configLock.lock()
        defer { configLock.unlock() }
        
        // Extract configuration values
        if let enabled = config["enabled"] as? Bool {
            configuration.setEnabled(enabled)
        }
        
        if let interval = config["monitoringInterval"] as? TimeInterval {
            if case .failure = configuration.setMonitoringInterval(interval) {
                reject(
                    "INVALID_INTERVAL",
                    "Invalid monitoring interval specified",
                    NSError(domain: kErrorDomain, code: -3)
                )
                return
            }
        }
        
        if let batteryOptimization = config["batteryOptimizationEnabled"] as? Bool {
            configuration.setBatteryOptimizationEnabled(batteryOptimization)
        }
        
        if let batteryThreshold = config["minimumBatteryLevel"] as? Double {
            if case .failure = configuration.setBatteryThreshold(batteryThreshold) {
                reject(
                    "INVALID_THRESHOLD",
                    "Invalid battery threshold specified",
                    NSError(domain: kErrorDomain, code: -4)
                )
                return
            }
        }
        
        // Update monitoring service with new configuration
        monitoringService.updateBatteryOptimization()
        
        resolve(["status": "updated"])
    }
    
    // MARK: - Private Methods
    @objc private func handleBatteryLevelChange(_ notification: Notification) {
        let batteryLevel = UIDevice.current.batteryLevel
        if batteryLevel <= configuration.minimumBatteryLevel {
            serialQueue.async { [weak self] in
                guard let self = self else { return }
                if self.monitoringService.isMonitoring {
                    self.monitoringService.stopMonitoring()
                    self.sendEvent(
                        withName: "onError",
                        body: ["code": "LOW_BATTERY", "message": "Monitoring stopped due to low battery"]
                    )
                }
            }
        }
    }
}

// MARK: - BackgroundTaskDelegate
extension BackgroundTaskModule: BackgroundTaskDelegate {
    public func backgroundTaskDidStart() {
        sendEvent(withName: "onMonitoringStateChange", body: ["isMonitoring": true])
    }
    
    public func backgroundTaskDidStop(_ error: Error?) {
        var body: [String: Any] = ["isMonitoring": false]
        if let error = error {
            body["error"] = error.localizedDescription
        }
        sendEvent(withName: "onMonitoringStateChange", body: body)
    }
    
    public func backgroundTaskDidDetectCry(
        _ audioData: Data,
        confidence: Double,
        timestamp: Date,
        metadata: [String: Any]
    ) {
        let eventBody: [String: Any] = [
            "confidence": confidence,
            "timestamp": timestamp.timeIntervalSince1970,
            "metadata": metadata
        ]
        sendEvent(withName: "onCryDetected", body: eventBody)
    }
}